caikit/core/model_management/model_trainer_base.py

```diff
@@ -0,0 +1,88 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A Trainer is responsible for managing execution of a training job for a given
+module class
+"""
+
+# Standard
+from enum import Enum
+from typing import Optional, Type
+import abc
+
+# Local
+from ..modules import ModuleBase
+from ..toolkit.factory import FactoryConstructible
+
+
+class ModelTrainerBase(FactoryConstructible):
+    __doc__ = __doc__
+
+    class TrainingStatus(Enum):
+        """A given training job must be in exactly one of these states at all
+        times
+        """
+
+        QUEUED = 1
+        RUNNING = 2
+        COMPLETED = 3
+        CANCELED = 4
+        ERRORED = 5
+
+    class ModelFutureBase(abc.ABC):
```

Comment:
nit: Sorry for being a bit pedantic about names, but would a more appropriate name for this class be: `TrainFutureBase` ? The reason being it exposes functions like `cancel` , `wait` etc, which would be semantically more related to "training future", also since its being returned as an output of `.train` function ?

Reply:
So my reason for `ModelFuture` instead of `TrainFuture` is that the thing that will be realized in the future is a `model` not a `training` (that's happening right now 😉). In general, my reading of the word "future" is that it's a "thing that will be true/available some time in the future," so it's really an implementation of [the proxy pattern](https://en.wikipedia.org/wiki/Proxy_pattern) where the `future` proxies for the thing that will eventually exist. The fact that it has `cancel`, `wait`, and `get_status` is the "interface" for being a `future`.

Reply:
make sense, its sort of a place-holder for what user might have gotten if it was a sync operation.